### PR TITLE
Media Modal: Fix invisible crop selection

### LINKED ISSUE #8398: Media Modal: Crop selection not always visible
## Bug Description

Originally reported [here](https://wordpress.org/support/topic/missing-preview-button-cropping-poster/), Lucky and I were able to reproduce this issue when trying to select a poster image.

In some cases the crop selection is not visible and trying to crop in that state results in an error.

Potentially useful files for debugging:

https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/media/controllers/cropper.js#L53-L72

https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/media/views/cropper.js#L82-L83

https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/_enqueues/vendor/imgareaselect/jquery.imgareaselect.js#L42

## Expected Behaviour

Cropping selection should always be visible.

## Steps to Reproduce

1. Make browser window small
2. Click to edit/add poster image
3. Select a large image that would need to be cropped
4. See no cropping selection
5. Click "Crop image" and get an error
6. Make window bigger
7. See cropping selection, drag it around if wanted
8. Click "Crop image" and see how image is sucessfully cropped

## Screenshots

Screen Shot 2021-07-20 at 10 01 59 AM

https://user-images.githubusercontent.com/841956/126442548-e7111450-a824-4569-9d33-f838aea46aa3.mov

## Additional Context

- Plugin Version: 1.9.0
- WordPress Version: 5.7
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

There is a issue with the imgareaselect library. Fix this by patching it locally and overriding and loading our version.  

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:
1. Open editor
1. Click to edit/add poster image
1. Make browser window small
1. Select a large image that would need to be cropped
1. See cropping selection
1. Click "Crop image" 

Try resizing window to different sizing while in cropping screen.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8398

### CODE REVIEW

File: includes/Admin/Crop_Script.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+use Google\Web_Stories\Infrastructure\Registerable;
+use Google\Web_Stories\Infrastructure\Service;
+use Google\Web_Stories\Traits\Screen;
+
+/**
+ * Class Crop Script
+ *
+ * @package Google\Web_Stories
+ */
+class Crop_Script implements Conditional, Service, Registerable {
```
Comment (spacedmonkey) [inline includes/Admin/Crop_Script.php]: @swissspidy Please help me with a better name here. Couldn't think of anything.

File: includes/Admin/Crop_Script.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @since 1.8.0
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed(): bool {
+		return is_admin() && ! wp_doing_ajax();
+	}
```
Comment (spacedmonkey) [inline includes/Admin/Crop_Script.php]: Make it conditional, why not.

File: includes/Admin/Crop_Script.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Runs on instantiation.
+	 *
+	 * @since 1.8.0
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_filter( 'script_loader_tag', [ $this, 'script_loader_tag' ], 10, 3 );
+	}
```
Comment (spacedmonkey) [inline includes/Admin/Crop_Script.php]: I looked at unregistering the script. However, the using this filter, means I don't have to register dependencies and means that is_edit_screen is usable.

File: includes/Admin/Crop_Script.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	 * @since 1.10.0
+	 *
+	 * @param string $tag    The `<script>` tag for the enqueued script.
+	 * @param string $handle The script's registered handle.
+	 * @param string $src    The script's source URL.
+	 *
+	 * @return string
+	 */
+	public function script_loader_tag( $tag, $handle, $src ) {
+		if ( self::SCRIPT_HANDLE !== $handle || ! $this->is_edit_screen() ) {
```
Comment (spacedmonkey) [inline includes/Admin/Crop_Script.php]: Only do the replacement in our editor.

File: includes/Assets.php [hunk 5]
```diff
@@ -81,7 +81,7 @@ public function get_base_url( string $path ): string {
 	 *
 	 * @return array Array containing combined contents of "<$handle>.asset.php" and "<$handle>.chunks.php".
 	 */
-	protected function get_asset_metadata( string $handle ): array {
+	public function get_asset_metadata( string $handle ): array {
```
Comment (spacedmonkey) [inline includes/Assets.php:84]: Needed so you can access this method in our new class.

File: packages/imgareaselect/src/index.js [hunk 6]
```diff
@@ -0,0 +1,1236 @@
...
+ * version 0.9.10-wp
+ *
+ * Copyright (c) 2008-2013 Michal Wojciechowski (odyniec.net)
+ *
+ * Dual licensed under the MIT (MIT-LICENSE.txt)
+ * and GPL (GPL-LICENSE.txt) licenses.
+ *
+ * https://github.com/odyniec/imgareaselect
+ *
+ */
```
Comment (spacedmonkey) [inline packages/imgareaselect/src/index.js:12]: Is there a copyright issue here?

File: packages/imgareaselect/src/index.js [hunk 7]
```diff
@@ -0,0 +1,1236 @@
...
+       * need to be rounded.
+       */
+      imgOfs = { left: round($img.offset().left), top: round($img.offset().top) };
+
+      /* Get image dimensions */
+      imgWidth = Math.ceil($img.innerWidth());
+      imgHeight = Math.ceil($img.innerHeight());
+
+      imgInnerWidth = Math.ceil($img.innerWidth());
+      imgInnerHeight = Math.ceil($img.innerHeight());
```
Comment (spacedmonkey) [inline packages/imgareaselect/src/index.js]: These are the patched lines

File: includes/Admin/Crop_Script.php (hunk 1 above)
Comment (swissspidy) [reply includes/Admin/Crop_Script.php]: `ImgAreaSelect_Patch`?

File: includes/Plugin.php [hunk 8]
```diff
@@ -84,6 +84,7 @@ class Plugin extends ServiceBasedPlugin {
 		'editor'                       => Admin\Editor::class,
 		'embed_shortcode'              => Shortcode\Embed_Shortcode::class,
 		'experiments'                  => Experiments::class,
+		'crop_script'                  => Admin\Crop_Script::class,
```
Comment (swissspidy) [inline includes/Plugin.php]: `crop_script` -> `imgareaselect_patch`?

File: packages/imgareaselect/README.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# imgareaselect
+
+ImgAreaSelect is a jQuery plugin for selecting a rectangular area of an image. This is a fork of the original jQuery [plugin](https://github.com/odyniec/imgareaselect).
```
Comment (swissspidy) [inline packages/imgareaselect/README.md]:
```suggestion
ImgAreaSelect is a jQuery plugin for selecting a rectangular area of an image. This is a fork of the original jQuery [plugin](https://github.com/odyniec/imgareaselect), as [patched by WordPress core](https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/_enqueues/vendor/imgareaselect/jquery.imgareaselect.js), with [one additional bugfix](https://github.com/odyniec/imgareaselect/issues/71#issuecomment-404979992).
```

File: webpack.config.cjs [hunk 10]
```diff
@@ -362,11 +362,26 @@ const storiesMCEButton = {
...
 
+const storiesImgareaselect = {
+  ...sharedConfig,
+  entry: {
+    imgareaselect: './packages/imgareaselect/src/index.js',
+  },
+  plugins: [
+    ...sharedConfig.plugins,
+    new WebpackBar({
+      name: 'WP imgareaselect',
```
Comment (swissspidy) [inline webpack.config.cjs]:
```suggestion
      name: 'WP ImgAreaSelect Patch',
```

File: packages/imgareaselect/src/index.js [hunk 11]
```diff
@@ -0,0 +1,1236 @@
...
+
+    /**
+     * Set plugin options
+     *
+     * @param newOptions
+     *            The new options object
+     */
+    function setOptions(newOptions) {
+      if (newOptions.parent)
+        ($parent = $(newOptions.parent)).append($box.add($outer));
```
Comment (swissspidy) [inline packages/imgareaselect/src/index.js]: GitHub reports a potential XSS vulnerability here, here's how we can fix it.

Not related to this original bug of course, but I also don't want to simply dismiss this issue here.

So let's just fix it.

```suggestion
        ($parent = $.find(newOptions.parent)).append($box.add($outer));
```

File: includes/Assets.php (hunk 5 above)
Comment (swissspidy) [reply includes/Assets.php:84]: That means we can update the `Assets` unit tests where this is called like this:

```php
$results = $this->call_private_method( $assets, 'get_asset_metadata', [ 'test_script' ] );
```

No longer need `call_private_method`

File: includes/Admin/Crop_Script.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	 * @param Assets $assets Assets instance.
+	 */
+	public function __construct( Assets $assets ) {
+		$this->assets = $assets;
+	}
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @since 1.8.0
```
Comment (swissspidy) [inline includes/Admin/Crop_Script.php]:
```suggestion
	 * @since 1.10.0
```

File: includes/Admin/Crop_Script.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed(): bool {
+		return is_admin() && ! wp_doing_ajax();
+	}
+
+	/**
+	 * Runs on instantiation.
+	 *
+	 * @since 1.8.0
```
Comment (swissspidy) [inline includes/Admin/Crop_Script.php]:
```suggestion
	 * @since 1.10.0
```

File: packages/imgareaselect/src/index.js (hunk 6 above)
Comment (swissspidy) [reply packages/imgareaselect/src/index.js:12]: Should be OK due to the dual licensing

**Review by swissspidy — APPROVED** [review]

**Comment by swissspidy** [comment]
Nice work! 🚀 

Just needs some QA instructions and then we can pass this on to Claudio.

**Comment by spacedmonkey** [comment]
> Just needs some QA instructions and then we can pass this on to Claudio.

Added

File: packages/imgareaselect/src/index.js [hunk 14]
```diff
@@ -0,0 +1,1241 @@
...
+    /**
+     * Set plugin options
+     *
+     * @param newOptions
+     *            The new options object
+     */
+    function setOptions(newOptions) {
+      if (newOptions.parent) {
+        // Patched XSS vulnerability.
+        ( $parent = $.find( newOptions.parent ) ).append( $box.add( $outer ) );
```
Comment (spacedmonkey) [inline packages/imgareaselect/src/index.js]: @swissspidy This line seems to be the problem on QA env. 
```
imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1 Uncaught TypeError: (intermediate value)(intermediate value).append is not a function
    at Ce (imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1)
    at ke (imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1)
    at new e.imgAreaSelect (imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1)
    at HTMLImageElement.<anonymous> (imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1)
    at Function.each (load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2)
    at s.fn.init.each (load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2)
    at s.fn.init.e.fn.imgAreaSelect (imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1)
    at n.onImageLoad (media-views.min.js?ver=5.7.2:2)
    at HTMLImageElement.dispatch (load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2)
    at HTMLImageElement.v.handle (load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2)
Ce @ imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1
ke @ imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1
e.imgAreaSelect @ imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1
(anonymous) @ imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1
each @ load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2
each @ load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2
e.fn.imgAreaSelect @ imgareaselect.js?ver=eaa352ba5401927608167fef838cec8a:1
onImageLoad @ media-views.min.js?ver=5.7.2:2
dispatch @ load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2
v.handle @ load-scripts.php?c=1&load[chunk_0]=jquery-core,jquery-migrate,utils&ver=5.7.2:2
```

I think we should revert this change and either leave it as it is or fix in another PR.

Comment (spacedmonkey) [reply packages/imgareaselect/src/index.js:918]: Couldn't replicate this locally

Comment (swissspidy) [reply packages/imgareaselect/src/index.js]: Looks like `newOptions.parent` can be a jQuery object, so we can write it like this:

```js
if (newOptions.parent) {
  $parent = newOptions.parent instanceOf $ ? newOptions.parent : $.find( newOptions.parent );
  $parent.append( $box.add( $outer ) );
}
```

Let's try that first.

Comment (swissspidy) [reply packages/imgareaselect/src/index.js]: Of course if this doesn't work, we can revert it. But I'd like to test it first on the QA site.

Comment (spacedmonkey) [reply packages/imgareaselect/src/index.js]: Where was the original report for this xss? I personally can't see the issue, the value submitted to the selector is a node not a string selector.

Comment (spacedmonkey) [reply packages/imgareaselect/src/index.js]: > Looks like `newOptions.parent` can be a jQuery object, so we can write it like this:
> 
> ```js
> if (newOptions.parent) {
>   $parent = newOptions.parent instanceOf $ ? newOptions.parent : $.find( newOptions.parent );
>   $parent.append( $box.add( $outer ) );
> }
> ```
> 
This didn't work. 

```
media-views.js?ver=5.8:6990 Uncaught TypeError: this.$image.imgAreaSelect is not a function
    at n.onImageLoad (media-views.js?ver=5.8:6990)
    at Zn (underscore.min.js?ver=1.8.3:2)
    at HTMLImageElement.<anonymous> (underscore.min.js?ver=1.8.3:2)
    at HTMLImageElement.<anonymous> (underscore.min.js?ver=1.8.3:2)
    at HTMLImageElement.dispatch (jquery.js?ver=3.6.0:5430)
    at HTMLImageElement.elemData.handle (jquery.js?ver=3.6.0:5234)
onImageLoad @ media-views.js?ver=5.8:6990
Zn @ underscore.min.js?ver=1.8.3:2
(anonymous) @ underscore.min.js?ver=1.8.3:2
(anonymous) @ underscore.min.js?ver=1.8.3:2
dispatch @ jquery.js?ver=3.6.0:5430
elemData.handle @ jquery.js?ver=3.6.0:5234
```
This does work 
```
$parent = newOptions.parent.append( $box.add( $outer ) );
```

**Review by swissspidy — APPROVED** [review]

File: packages/imgareaselect/src/index.js (hunk 14 above)
Comment (swissspidy) [reply packages/imgareaselect/src/index.js]: > Where was the original report for this xss? I personally can't see the issue, the value submitted to the selector is a node not a string selector.

See https://github.com/google/web-stories-wp/pull/8501#discussion_r679007074

It's GitHub security check. Only visible to admins.

And it's only a _potential_ vulnerability known to jQuery plugins in general, where the options are passed to `$()` unsanitized.

**Pull request merged**